052-prometheus-metrics-reporter.md

```diff
@@ -0,0 +1,171 @@
+- Yammer, `com.yammer.metrics.metrics-core`. This library is the predecessor of [Dropwizard](https://metrics.dropwizard.io/). We’ll refer to metrics created via this as _YammerMetrics_. This is only used on the broker side.
+
+For each type of metrics, Kafka exposes a reporter interface to expose metrics to monitoring systems. KafkaMetrics use [org.apache.kafka.common.metrics.MetricsReporter](https://kafka.apache.org/35/javadoc/org/apache/kafka/common/metrics/MetricsReporter.html) and YammerMetrics use `kafka.metrics.KafkaMetricsReporter` which is not officially part of the public API. Kafka has built-in metrics reporter implementations for JMX for both types.
+
+At the moment Strimzi relies on these default JMX reporters and uses [jmx_exporter](https://github.com/prometheus/jmx_exporter) which is a Java agent that retrieves metrics via JMX and exposes them over an HTTP endpoint in the prometheus format. Then Prometheus is configured to scrape that endpoint to retrieve the Kafka metrics.
+
+![Current Situation](images/052-current.png)
+
+- `org.apache.kafka.common.metrics.JmxReporter` is the reporter implementation for the Kafka Metrics.
+- `org.apache.kafka.server.metrics.FilteringJmxReporter` is the reporter implementation for the Yammer Metrics. It’s named `FilteringJmxReporter` because it extends the JmxReporter class from the Yammer library and adds an option to select the metrics to report.
+
+
+## Motivation
+
+I propose updating the metrics collection pipeline for the following reasons:
+
+1. The current metrics collection pipeline is pretty convoluted. We have metrics reporters first exposing metrics via JMX before using a Java agent to expose them again via HTTP. Using metrics reporters to directly expose metrics to Prometheus would significantly simplify it by removing JMX from the picture and removing jmx_exporter.
+1. Each component along the pipeline has its own configurations and specificities. For example, since Kafka 3.4.0, it’s possible to disable the JmxReporter for KafkaMetrics by setting [auto.include.jmx.reporter](https://kafka.apache.org/documentation/#brokerconfigs_auto.include.jmx.reporter) to `false`. It is currently not possible to disable `FilteringJmxReporter` (it’s something I plan to address in the near future). Also both jmx_exporter and the built-in reporter allow selecting metrics to collect. 
+1. The jmx_exporter Java agent supports complex metrics mapping rules. These rules allow renaming metrics and it can make it hard to investigate issues as metrics could have different names. This prevents using metrics like an API, whether this is to build grafana dashboards or for the operator to rely on metrics.
+1. Due to the complex rules, it performs badly when there’s a very large number of metrics due to a lot of topic/partitions.
+1. Recently jmx_exporter was impacted by CVEs in its dependencies (CVE-2022-38752, CVE-2022-25857).
```

Comment:
Should this be a numbered list and not just `1.`, `1.`, `1.` ...?

Reply:
In markdown you don't need to actually properly number each items. This renders correctly as 1., 2., 3., etc. See https://github.com/strimzi/proposals/blob/219127ab4d7a3f242fb941ee81336b60b41f2961/052-prometheus-metrics-reporter.md

Reply:
I guess that beats the purpose of Markdown as something that is easily readable without any editor/viewer even as a plain text ;-)

Reply:
Makes the review comments funny as well ... for example `the fifth option 1. does not make sense` :-D